{% extends "base.html" %} {% block title %} COMET Home{% endblock %} {% block content%}
<div class="container">
    <h1>About</h1>
    <hr>
    <h2><strong>Purpose</strong></h2>
    <p>The goal of COMET is to determine the effect of SNPs using protein-protein interactions (PPIs), specifically in
        the calculation of the ∆∆G value.</p>
    <h2><strong>Understanding the SNP results</strong></h2>
    <h3>SNP gene, chromosome and genotype</h3>
    <p>Single nucleotide polymorphisms (SNPs) are the most common type of genetic variation. SNPs are locations in the
        DNA where the base nucleotide (A, T, C or G) tend to vary between different people.</p>
    <p>The genotype is the nucleotide(s) that the genetic test determined. Autosomal chromosomes (chromosomes 1-22) come
        in pairs so there are two nucleotides for the reported genotype. The genotype of the sex chromosomes depends
        on the sex of the person genotyped.</p>
    <h3>Clinical significance</h3>
    <p>Clinical significance is obtained from dbSNP. Example outputs include benign, likely benign, uncertain
        significance, likely pathogenic, pathogenic, drug response, risk factor, etc. For more information please visit
        NCBI's ClinVar documentation <a href="https://www.ncbi.nlm.nih.gov/clinvar/docs/clinsig/">(ClinVar documentation
            link)</a>.</p>
    <h3>Related conditions and diseases</h3>
    <p>Related conditions are obtained from ClinVar and related diseases are obtained from LitVar. The ClinVar
        conditions include the number of times the condition appears in the database whereas the LitVar diseases
        include the number of published literature about the disease and the given SNP.</p>
    <h3>Substitutions</h3>
    <p>For a given SNP, there can be multiple amino acid subsitutions and/or nucleotide subsitutions. Multiple
        substitutions may be due to the protein having multiple isomers (protein variants) as a result of alternative
        splicing.</p>
    <h3>Frequencies</h3>
    <p>The frequency of each SNP is compared to The 1000 Genomes Project and the HapMap Project when available.</p>
    <h3>Change in binding affinity (∆∆G)</h3>
    <p>The ∆∆G value is predicted using four programs (SAAMBE-3D, I-Mutant2.0 (sequence and structure), PANDA and UEP).
        The stability consensus is determined based on the number of programs that indicate an increase in stability
        verusus a decrease in stability. </p>
    <p>Regarding the interpretation of ∆∆G values, the effect of the mutation is determined by the sign of the ∆∆G
        value. Positive ∆∆G values for I-Mutant2.0 and PANDA indicate an increase in stability whereas for SAAMBE-3D and
        UEP indicate a decrease in stability. Thus, negative ∆∆G values for I-Mutant2.0 and PANDA indicate a decrease
        in stability whereas for SAAMBE-3D and UEP indicate an increase in stability.</p>
    <p>These programs accept the protein sequence (obtained from Uniprot) and/or the protein (PDB) structure (obtained
        from RCSB). Sometimes the programs may not produce an output. If there is no PDB structure, SAAMBE-3D,
        I-Mutant2.0 (structure) and UEP cannot calculate the ∆∆G value. If there is no sequence, I-Mutant2.0 (structure)
        and PANDA cannot calculate the ∆∆G value. Additionally, UEP only evaluates mutations located in highly-packed
        interface positions. UEP and and PANDA also rely on an explicitly stated secondary protein to calculate the ∆∆G
        value.</p>

    <h3>Protein Viewer</h3>
    <p>The protein viewer uses the NGL viewer to display the PDB structure used to calculate the ∆∆G value.</p>
    <h2><strong>Uploading raw genotype data</strong></h2>
    <p>If you do not have a raw genotype file or prefer to not use your own, you can use the browse SNP option.
        Alternatively, you can freely download raw genotype files from OpenSnp. (<a href="https://opensnp.org/">OpenSNP
            Link</a>).
        All raw genotype data is automatically deleted when the End Session button is clicked. If you forget to
        click
        this button, the data will be automatically deleted after 24 hours.</p>
</div>
{% endblock %}