{% extends "base.html" %} {% block title %}{{snp}} Results{% endblock %} {% block
    content %}

<head>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/Chart.js/2.7.2/Chart.min.js"></script>
  <script type="text/javascript" src="http://ajax.googleapis.com/ajax/libs/jquery/1.6.2/jquery.min.js"></script>
  <link href="https://cdn.jsdelivr.net/npm/bootstrap@5.0.0-beta1/dist/css/bootstrap.min.css" rel="stylesheet"
    integrity="sha384-giJF6kkoqNQ00vy+HMDP7azOuL0xtbfIcaT9wjKHr8RbDVddVHyTfAAsrekwKmP1" crossorigin="anonymous">

</head>

<body>
  <div class="container">
    <!-- <div class id="title"> -->
      <div class="row">
        <div class="col">
          <h1>{{snp}} Results</h1>
          <div role="group" aria-label="Navigation buttons">
            <a class="btn btn-secondary jumpto" href="/select">Back to table</a>
            <a class="btn btn-secondary jumpto" href="/prevsnp">Previous SNP</a>
            <a class="btn btn-secondary jumpto" href="/nextsnp">Next SNP</a>
          </div>
        </div>
        <div class="col jumpto">
          <p></p>
          <p>Jump to:</p>
          <p><a href="#sub-freq">Substitutions and frequencies</a></p>
          <p><a href="#bind-affinity">Change in binding affinity (∆∆G)</a></p>
          <p><a href="#protein-viewer">Protein viewer</a></p>
        </div>

    </div>
    
    <p></p>
    <div class="row">
      <div class="col-sm snp-info">
        <p>Gene: {{gene}}</p>
        <P>Chromosome: {{chr}}</P>
        <p>Genotype: {{genotype}}</p>
      </div>
      <div class="col-sm snp-info">
        <p>Clinical significance:</p>
        {% for item in clin %}
        <ul>
          <li>{{item}}</li>
        </ul>
        {% endfor %}
      </div>
    </div>

    <h3 class="result-titles" id="sub-freq">Substitutions and frequencies</h3>
    <div class="snp-info">
      <p>For a given SNP, there can be multiple amino acid subsitutions and/or nucleotide subsitutions. Multiple
        substitutions may be due to the protein having multiple isomers (forms).</p>
      <p>These polar area charts (which are similar to a pie chart but shows fractions in terms of area rather than
        angle) represent the nucleotide and/or amino acid substiutions that are documented for this SNP. The larger
        values indicate the subsitutions that have a greater association with the given SNP.</p>
    </div>
    <div class="left-box"><canvas id="chart1" class="left-graph" width="20" height="10" ></canvas></div>
    <div class="right-box"><canvas id="chart2" class="right-graph" width="20" height="10"></canvas></div>
    <p>The nucleotide substitutions (with frequency) are
      {% for i in range(0, len(sorted_nuclist[0])) %}
      {% if i == len(sorted_nuclist[0]) - 1 %}
      {{sorted_nuclist[0][i]}} ({{sorted_nuclist[1][i]}}).
      {% else %}
      {{sorted_nuclist[0][i]}} ({{sorted_nuclist[1][i]}}),
      {% endif %}
      {% endfor %}
      The amino acid substitutions (with frequency) are
      {% for i in range(0, len(sorted_aalist[0])) %}
      {% if i == len(sorted_nuclist[0]) - 1 %}
      {{sorted_aalist[0][i]}} ({{sorted_aalist[1][i]}}).
      {% else %}
      {{sorted_aalist[0][i]}} ({{sorted_aalist[1][i]}}),
      {% endif %}
      {% endfor %}</p>

    <div class="snp-info">
      <p>For a given SNP, the genotype can be compared to Genome Wide Studies such as The 1000 Genomes Project or the
        HapMap Project.</p>
      <p>These pie charts represent the frequency of the nucleotide corresponding to the SNP compared to the frequency
        of all other nucleotides. The smaller percentage indicates rarer SNPs.</p>
    </div>
    {% if freq1000g != 'N/A' and freqhapmap != 'N/A' %}
    <div class="left-box"><canvas id="pie1" width="20" height="10" class="left-graph"></canvas></div>
    <div class="right-box"><canvas id="pie2" width="20" height="10" class="right-graph"></canvas></div>
    <p>From the 1000 Genomes Project, the {{ freq1000g[0] }} nuclotide has a frequency of {{freq1000g[3:]}}
      and from the HapMap Project, the {{ freqhapmap[0] }} nuclotide has a frequency of {{ freqhapmap[3:] }}.</p>

    {% else %}
    <p>There are no nucleotide frequencies for this SNP from the 1000 Genomes Project or the HapMap Project.</p>
    {% endif %}

    <script>
      var chart1 = new Chart(document.getElementById("chart1"), {
        type: 'polarArea',
        data: {
          labels: '{{sorted_nuclist[0]| tojson}}'.slice(1, -1).split(', '),
          datasets: [{
            backgroundColor: ["#351f39", "#726a95", "#719fb0", "#a0c1b8", "#c7cfb7", "#9dad7f", "#557174",
              "#889bae"
            ],
            data: '{{sorted_nuclist[1]| tojson}}'.slice(1, -1).split(', ')
          }]
        },
        options: {
          title: {
            display: true,
            text: 'Nucleotide Substitutions'
          },
          legend: {
            position: 'left'
          }
        }
      });
      var aa_val = '{{sorted_aalist[1]| tojson}}'.slice(1, -1).split(', ')
      var aa_label = '{{sorted_aalist[0]| tojson}}'.slice(1, -1).split(', ')
      if ('{{aa1}}' == 'N/A') {
        aa_val = [0];
        aa_label = ['N/A'];
      }
      var chart2 = new Chart(document.getElementById("chart2"), {
        type: 'polarArea',
        data: {
          labels: aa_label,
          datasets: [{
            backgroundColor: ["#351f39", "#726a95", "#719fb0", "#a0c1b8", "#c7cfb7", "#9dad7f", "#557174",
              "#889bae"
            ],
            data: aa_val
          }]
        },
        options: {
          title: {
            display: true,
            text: 'Amino acid Substitutions'
          },
          legend: {
            position: 'right'
          }
        }
      });
      var fraction1 = parseFloat('{{ freq1000g[3:] }}')
      var fraction2 = parseFloat('{{ freqhapmap[3:] }}')
      var colors = ['#5d1451', '#2f416d'];
      var pie1 = new Chart(document.getElementById("pie1"), {
        type: 'pie',
        data: {
          labels: ['{{ freq1000g[0] }}', "Other"],
          datasets: [{
            backgroundColor: colors,
            data: [fraction1, 1 - fraction1]
          }]
        },
        options: {
          title: {
            display: true,
            text: '1000 Genomes Project'
          },
          legend: {
            position: 'left'
          }
        }
      });
      var pie2 = new Chart(document.getElementById("pie2"), {
        type: 'pie',
        data: {
          labels: ['{{ freqhapmap[0] }}', "Other"],
          datasets: [{
            backgroundColor: colors,
            data: [fraction2, 1 - fraction2]
          }]
        },
        options: {
          title: {
            display: true,
            text: 'HapMap Project'
          },
          legend: {
            position: 'right'
          }
        }
      });
    </script>

    <h3 class="result-titles" id="bind-affinity">Change in binding affinity (∆∆G)</h3>
    <div class="snp-info">
      <p>The ∆∆G value can only be determined in there is a known amino acid substitution. The ∆∆G value can be
        calculated from wild -> mutated or mutated -> wild.</p>
      <p>UEP only evaluates mutations located in highly-packed interface positions. SAAMBE-3D and I-Mutant2.0 Structure
        only calcuate ∆∆G if the position is present on the pdb structure.</p>
    </div>
    {% if aa1 != 'N/A' %}
    <form action="#" method="post">
      <input type="submit" name="ddgcalc" value="wild" onclick="$('#loading').show();">
      <input type="submit" name="ddgcalc" value="mutated" onclick="$('#loading').show();">
    </form>
    <div id="loading">
      <p>Loading ... </p>
      <!-- <img src="static/loadingimage.gif" alt="loading"> -->
    </div>
    {% endif %}
    {% if ddgresults %}
    <p>Calculated with {{pdbselect}} and {{aa1}}</p>

    <table id="example" class="display">
      <tr>
        <th>Program</th>
        <th>∆∆G (kcal/mol)</th>
        <th>Effect</th>
      </tr>
      {% for i in range(0, len(ddgresults)-1) %}
      <tr>
        <td>{{ddgresults[i][0]}}</td>
        <td>{{ddgresults[i][1]}}</td>
        <td>{{ddgresults[i][2]}}</td>
      </tr>
      {% endfor %}
      <tr>
        <th>{{ddgresults[len(ddgresults)-1][0]}}</th>
        <th></th>
        <th>{{ddgresults[len(ddgresults)-1][2]}}</th>
      </tr>
    </table>
    {% endif %}
    {% if pdb != 'N/A' %}

    <h3 class="result-titles" id="protein-viewer">Protein viewer </h3>
    <div class="snp-info">
      <p>To view protein, calculate change in binding affinity. </p>

      {% if ddgresults %}
      <p>The specific areas (if any) that the {{gene}} protein is affected by {{snp}} in {{pdbselect}} is shown in
        purple.
      </p>
      {% endif %}
    </div>

    <div>
      <form action="#" method="post">
        {% for i in pdb %}
        <input type="submit" name="pdbselect" value={{i}}>
        {% endfor %}
      </form>

      <script src="https://cdn.rawgit.com/arose/ngl/v0.10.4-1/dist/ngl.js "></script>

      <script>
        var pdbfile = '{{ pdbselect }}';
        var rscb = "rcsb://";
        var input = rscb.concat(pdbfile);
        var chain = ':{{chain}}';
        document.addEventListener("DOMContentLoaded", function () {
          var stage = new NGL.Stage("viewport", {
            backgroundColor: 'white'
          });
          var schemeId = NGL.ColormakerRegistry.addSelectionScheme([
            ["purple", chain],
            ["grey", "*"]
          ])
          stage.loadFile(input).then(function (o) {
            o.addRepresentation("ribbon", {
              colorScheme: schemeId
            });
            o.autoView();
          });
        });
      </script>
      <div id="viewport"></div>
    </div>
    {% endif %}

  </div>
</body>

{% endblock %}