{% extends "base.html" %} {% block title %}{{snp}} Results{% endblock %} {% block
    content %}

<head>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/Chart.js/2.7.2/Chart.min.js"></script>
  <script type="text/javascript" src="http://ajax.googleapis.com/ajax/libs/jquery/1.6.2/jquery.min.js"></script>
  <link href="https://cdn.jsdelivr.net/npm/bootstrap@5.0.0-beta1/dist/css/bootstrap.min.css" rel="stylesheet"
    integrity="sha384-giJF6kkoqNQ00vy+HMDP7azOuL0xtbfIcaT9wjKHr8RbDVddVHyTfAAsrekwKmP1" crossorigin="anonymous">
  <link rel="stylesheet" href="{{ url_for('static', filename='style.css') }}">
  <link rel="shortcut icon" href="{{ url_for('static', filename='comet.ico') }}">
  <title>Results</title>
</head>

<body>
  <div class="container">
    <div class="row">
      <div class="col-sm">
        <h1>{{snp}} Results</h1>
        <div role="group" aria-label="Navigation buttons">
          <a class="btn btn-secondary jumpto" href="/select">Back to table</a>
          <a class="btn btn-secondary jumpto" href="/prevsnp">Previous SNP</a>
          <a class="btn btn-secondary jumpto" href="/nextsnp">Next SNP</a>
          <a class="btn btn-outline-secondary jumpto" onclick="printDiv()">Print SNP info</a>
        </div>
      </div>
      <div class="col jumpto">
        <p></p>
        <p>On this page:</p>
        {% if condition[1] != "N/A" %}
        <a href="#conds">Related conditions and diseases</a><br>
        {% endif %}
        <a href="#sub">Substitutions</a><br>
        {% if freq1000g != 'N/A' or freqhapmap != 'N/A' %}
        {% endif %}
        <a href="#freq">Frequencies</a><br>
        <a href="#bind-affinity">Change in binding affinity (∆∆G)</a><br>
        {% if pdb != 'N/A' %}
        <a href="#protein-viewer">Protein viewer</a><br>
        {% endif %}
      </div>
    </div>
    <div id="printme">
      <h2 class="result-titles">{{snp}}</h2>
      <hr>
      <div class="row">
        <div class="col">
          <p><strong>Gene: </strong>{{gene}}</p>
          <p></p>
          <p><strong>Chromosome: </strong>{{chr}}</p>
          {% if genotype != "N/A" and genotype != "undefined" %}
          <p></p>
          <p><strong>Your Genotype: </strong>{{genotype}}</p>

          {% endif %}

        </div>
        <div class="col">
          <strong>Clinical significance:</strong>
          {% for i in range(0, len(clin)) %}
          {% if i == len(clin) - 1 %}{{clin[i]}}.
          {% else %}{{clin[i]}},
          {% endif %}
          {% endfor %}
          <p></p>
          <p></p>
          <p><strong>Summary: </strong>{% if genotype != "N/A" and genotype != "undefined" %}Your genotype contains
            {{summary[0]}} nucleotide substitution(s) associated with
            this SNP. {% endif %} {% if summary[1] != -1 and freq1000g != 'N/A' %} The {{freq1000g[0]}} nucleotide is
            present in about
            {{summary[1]}}% of people. {% endif %}{% if ddgresults %}
            {% if ddgresults[len(ddgresults)-1][2] == "Decrease in stability" %} The amino acid
            substitution associated with this SNP causes a decrease in stability.
            {% endif %}{% if ddgresults[len(ddgresults)-1][2] == "Increase in stability" %}The amino acid substitution
            associated with this SNP causes an increase in
            stability.{% endif %}{% if ddgresults[len(ddgresults)-1][2] == "No change in stability" %}The amino acid
            substitution associated with this SNP has an inconclusive change in stability.{% endif %}</p>
          {% endif %}

        </div>
        <!-- <div class="w-100"></div> -->
      </div>
    </div>
    <script>
      function printDiv() {
        var divContents = document.getElementById("printme").innerHTML;
        var divContents2 = document.getElementById("printme2").innerHTML;
        var a = window.open('', '', 'height=500, width=500');
        a.document.write('<html>');
        a.document.write('<body>');
        a.document.write(divContents);
        a.document.write(divContents2);
        a.document.write('</body></html>');
        a.document.close();
        a.print();
      }
    </script>
    {% if condition[1] != "N/A" or len(condition[0]) != 0 %}
    <h2 class="result-titles" id="conds">Related conditions and diseases</h2>
    <hr>
    {% endif %}
    <div class="row">
      <div class="col">
        {% if condition[1] != "N/A" %}
        <canvas id="litvar"></canvas>
        {% endif %}
      </div>

      <div class="col">
        <div id="printme2">
          {% if len(condition[0]) != 0 %}
          <strong>ClinVar related conditions:</strong>
          {% for i in range(0, len(condition[0])) %}
          {% if i == len(condition[0]) - 1 %}{{condition[0][i]}}.
          {% else %}{{condition[0][i]}},
          {% endif %}
          {% endfor %}
          {% endif %}
          {% if condition[1] != "N/A" %}
          <p></p>
          <p></p>
          <strong>LitVar related diseases:</strong>
          {% for i in range(0, len(condition[1])) %}
          {% if i == len(condition[1]) - 1 %}
          {{condition[1][i]}} ({{condition[2][i]}}).
          {% else %}
          {{condition[1][i]}} ({{condition[2][i]}}),
          {% endif %}
          {% endfor %}
          {% endif %}
        </div>
      </div>
    </div>
    <script>
      var litvar = new Chart(document.getElementById("litvar"), {
        type: 'pie',
        data: {
          labels: '{{condition[1]| tojson}}'.slice(1, -1).split(', '),
          datasets: [{
            backgroundColor: ['#97c1a9', '#c6dbda', "#D4D479", "#CCE2CB", "#FFBF91", "#FFFFB5", "#FCB9AA",
              "#ECD5E3"
            ],
            data: '{{condition[2]| tojson}}'.slice(1, -1).split(', ')
          }]
        },
        options: {
          title: {
            display: true,
            text: 'LitVar published literature per related disease'
          },
          legend: {
            position: 'left'
          }
        }
      });
    </script>

    <h2 class="result-titles" id="sub">Substitutions</h2>
    <hr>
    <p class="grey-border">These pie charts represent the nucleotide and/or amino acid substiutions that are
      documented for this SNP.
      The larger values indicate the subsitutions that are more commonly associated with the given SNP.</p>

    <div class="left-box"><canvas id="chart1" class="left-graph" width="20" height="10"></canvas></div>
    <div class="right-box"><canvas id="chart2" class="right-graph" width="20" height="10"></canvas></div>

    <p>The nucleotide substitutions (with frequency) are
      {% for i in range(0, len(sorted_nuclist[0])) %}
      {% if i == len(sorted_nuclist[0]) - 1 %}
      {{sorted_nuclist[0][i]}} ({{sorted_nuclist[1][i]}}).
      {% else %}
      {{sorted_nuclist[0][i]}} ({{sorted_nuclist[1][i]}}),
      {% endif %}
      {% endfor %}
      {% if aa1 == 'N/A' %}
      There are no known amino acid substitutions.
      {% else %}
      The amino acid substitutions (with frequency) are
      {% for i in range(0, len(sorted_aalist[0])) %}
      {% if i == len(sorted_aalist[0]) - 1 %}
      {{sorted_aalist[0][i]}} ({{sorted_aalist[1][i]}}).
      {% else %}
      {{sorted_aalist[0][i]}} ({{sorted_aalist[1][i]}}),
      {% endif %}
      {% endfor %}
      {% endif %}
    </p>
    {% if freq1000g != 'N/A' or freqhapmap != 'N/A' %}
    <h2 class="result-titles" id="freq">Frequencies</h2>
    <hr>
    <p class="grey-border">These pie charts represent the frequency of the nucleotide corresponding to the SNP
      compared to the frequency
      of all other nucleotides. The smaller percentage indicates rarer SNPs.</p>
    <div class="left-box">
      <canvas id="pie1" class="left-graph" width="20" height="10"></canvas>
    </div>

    <div class="right-box">
      <canvas id="pie2" class="right-graph" width="20" height="10"></canvas>
    </div>
    <br>
    <p>{% if freq1000g != 'N/A' %}The {{ freq1000g[0]}} nucleotide has a frequency of
      {{round(float(freq1000g[3:])*100,2)}}% in the 1000
      Genomes Project. {% endif %}{% if freqhapmap != 'N/A' %}The {{ freqhapmap[0] }} nucleotide has a
      frequency of {{ round(float(freqhapmap[3:]),2) *100}}% in the HapMap Project.{% endif %}</p>
    {% else %}
    <p>There are no nucleotide frequencies for this SNP from the 1000 Genomes Project or the HapMap Project.</p>
    {% endif %}
    <script>
      var chart1 = new Chart(document.getElementById("chart1"), {
        type: 'pie',
        data: {
          labels: '{{sorted_nuclist[0]| tojson}}'.slice(1, -1).split(', '),
          datasets: [{
            backgroundColor: ['#97c1a9', '#c6dbda', "#D4D479", "#CCE2CB", "#FFBF91", "#FFFFB5", "#FCB9AA",
              "#ECD5E3", "#719fb0", "#351f39", "#c7cfb7", "#9dad7f", "#557174",
              "#889bae"
            ],
            data: '{{sorted_nuclist[1]| tojson}}'.slice(1, -1).split(', ')
          }]
        },
        options: {
          title: {
            display: true,
            text: 'Nucleotide Substitutions'
          },
          legend: {
            position: 'left'
          }
        }
      });
      var aa_val = '{{sorted_aalist[1]| tojson}}'.slice(1, -1).split(', ')
      var aa_label = '{{sorted_aalist[0]| tojson}}'.slice(1, -1).split(', ')
      if ('{{aa1}}' == 'N/A') {
        aa_val = [0];
        aa_label = ['N/A'];
      }
      var chart2 = new Chart(document.getElementById("chart2"), {
        type: 'pie',
        data: {
          labels: aa_label,
          datasets: [{
            backgroundColor: ['#97c1a9', '#c6dbda', "#D4D479", "#CCE2CB", "#FFBF91", "#FFFFB5", "#FCB9AA",
              "#ECD5E3", "#719fb0", "#a0c1b8", "#c7cfb7", "#9dad7f", "#557174",
              "#889bae"
            ],
            data: aa_val
          }]
        },
        options: {
          title: {
            display: true,
            text: 'Amino acid Substitutions'
          },
          legend: {
            position: 'right'
          }
        }
      });
      var fraction1 = parseFloat('{{ freq1000g[3:] }}')
      var fraction2 = parseFloat('{{ freqhapmap[3:] }}')
      var colors = ['#97c1a9', '#c6dbda'];
      var pie1 = new Chart(document.getElementById("pie1"), {
        type: 'pie',
        data: {
          labels: ['{{ freq1000g[0] }}', "Other"],
          datasets: [{
            backgroundColor: colors,
            data: [fraction1, 1 - fraction1]
          }]
        },
        options: {
          title: {
            display: true,
            text: '1000 Genomes Project'
          },
          legend: {
            position: 'left'
          }
        }
      });
      var pie2 = new Chart(document.getElementById("pie2"), {
        type: 'pie',
        data: {
          labels: ['{{ freqhapmap[0] }}', "Other"],
          datasets: [{
            backgroundColor: colors,
            data: [fraction2, 1 - fraction2]
          }]
        },
        options: {
          title: {
            display: true,
            text: 'HapMap Project'
          },
          legend: {
            position: 'right'
          }
        }
      });
    </script>

    <h2 class="result-titles" id="bind-affinity">Change in binding affinity (∆∆G)</h2>
    <hr>
    {% if aa1 != 'N/A' %}

    {% if ddgresults%}
    {% if pdb != "N/A" %}
    <p class="grey-border">Recalculate ∆∆G value with a different PDB structure or protein (if available). Only UEP
      and
      PANDA rely on
      an explicitly stated secondary protein.</p>
    <form action="#" method="post">
      <div>
        {% for i in pdb %}
        {% if i == pdbselect %}
        <input type="radio" id={{i}} name="pdbselect" value={{i}} checked>
        <label for={{i}}>{{i}}</label>
        {% if len(chain[2]) > 1 %}
        (
        {% for j in chain[2] %}
        {% if j == chain[3] %}
        <input type="radio" id={{i}} name="protselect" value={{j}} checked>
        {% else %}
        <input type="radio" id={{i}} name="protselect" value={{j}}>
        {% endif %}
        <label for={{j}}>{{j}}</label>
        {% endfor %}
        )
        {% endif %}
        {% else %}
        <input type="radio" id={{i}} name="pdbselect" value={{i}}>
        <label for={{i}}>{{i}}</label>
        {% endif %}
        {% endfor %}
      </div>
      <br>
      <div>
        <button type="submit" class="btn btn-primary" onclick="$('#loading').show();">Recalculate ∆∆G</button>
      </div>
    </form>
    {% endif %}
    {% else %}
    <p class="grey-border">Press the "Calculate ∆∆G" button to calculate the ∆∆G value using each predictor.</p>
    <form action="#" method="post">
      <input type="submit" class="btn btn-primary" name="ddgcalc" value="Calculate ∆∆G" onclick="$('#loading').show();">
    </form>
    {% endif %}
    <div id="loading">
      <p>Loading ... </p>
    </div>
    {% else %}
    <p class="grey-border">The ∆∆G value cannot be calculated as there are no known amino acid substitutions. </p>
    {% endif %}
    {% if ddgresults %}
    <br>
    <table id="results" class="display">
      {% if pdb != "N/A" %}
      <caption>∆∆G values calculated with {{pdbselect}} and {{aa1}}.</caption>
      {% else %}
      <caption>∆∆G values calculated with {{aa1}} on the sequence (no PDB structure available).</caption>
      {% endif %}
      <tr>
        <th id="program">Program</th>
        <th id="deltadeltaG">∆∆G (kcal/mol)</th>
        <th id="effect">Effect</th>
      </tr>
      {% for i in range(0, len(ddgresults)-1) %}
      <tr>
        <td headers="program {{ddgresults[i][0]}}">{{ddgresults[i][0]}}</td>
        <td headers="deltadeltaG {{ddgresults[i][0]}}">{{ddgresults[i][1]}}</td>
        <td headers="effect {{ddgresults[i][0]}}">{{ddgresults[i][2]}}</td>
      </tr>
      {% endfor %}
      <tr>
        <th id="conclusion">Stability Consensus</th>
        <th id="space">-</th>
        <th id="consensus">{{ddgresults[len(ddgresults)-1][2]}}</th>
      </tr>
    </table>
    <br>
    <p>Note: UEP only evaluates mutations located in highly-packed interface positions. SAAMBE-3D and I-Mutant2.0
      Structure only calculate ∆∆G if the position is present on the pdb structure. PANDA and UEP only calculate ∆∆G
      if
      the secondary sequence is available.</p>
    {% endif %}
    {% if pdb != 'N/A' %}

    <h2 class="result-titles" id="protein-viewer">Protein viewer </h2>
    <hr>
    {% if ddgresults and chain[0] != "*" %}
    {% if chain[3] != "N/A" %}
    <p class="grey-border">The {{pdbselect}} structure. The {{gene}} protein is shown in purple and the interacting
      protein ({{chain[3]}}) is shown in blue. </p>
    {% else %}
    <p class="grey-border">The {{pdbselect}} structure. The {{gene}} protein is shown in purple. </p>
    {% endif %}
    <div>
      <script src="https://cdn.rawgit.com/arose/ngl/v0.10.4-1/dist/ngl.js "></script>
      <script>
        var pdbfile = '{{ pdbselect }}';
        var rscb = "rcsb://";
        var input = rscb.concat(pdbfile);
        var chain1 = ':{{chain[0]}}';
        var chain2 = ':{{chain[1]}}';
        document.addEventListener("DOMContentLoaded", function () {
          var stage = new NGL.Stage("viewport", {
            backgroundColor: 'white'
          });
          var schemeId = NGL.ColormakerRegistry.addSelectionScheme([
            ["purple", chain1],
            ["blue", chain2],
            ["grey", "*"]
          ])
          stage.loadFile(input).then(function (o) {
            o.addRepresentation("ribbon", {
              colorScheme: schemeId
            });
            o.autoView();
          });
        });
      </script>
      <div id="viewport"></div>
    </div>
    {% else %}
    <p class="grey-border">Calculate (or recalculate) ∆∆G to view protein structure.</p>
    {% endif %}
    {% endif %}
  </div>
</body>
{% endblock %}