{% extends "base.html" %} {% block title %}{{snp}} Results{% endblock %} {% block
    content %}

<head>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/Chart.js/2.7.2/Chart.min.js"></script>
  <script type="text/javascript" src="http://ajax.googleapis.com/ajax/libs/jquery/1.6.2/jquery.min.js"></script>
  <link href="https://cdn.jsdelivr.net/npm/bootstrap@5.0.0-beta1/dist/css/bootstrap.min.css" rel="stylesheet"
    integrity="sha384-giJF6kkoqNQ00vy+HMDP7azOuL0xtbfIcaT9wjKHr8RbDVddVHyTfAAsrekwKmP1" crossorigin="anonymous">
  <link rel="stylesheet" href="{{ url_for('static', filename='style.css') }}">
  <link rel="shortcut icon" href="{{ url_for('static', filename='comet.ico') }}">
  <title>Results</title>
</head>

<body>
  <div class="container">
    <div class="row">
      <div class="col-sm">
        <h1>{{snp}} Results</h1>
        <div role="group" aria-label="Navigation buttons">
          <a class="btn btn-secondary jumpto" href="/select">Back to table</a>
          <a class="btn btn-secondary jumpto" href="/prevsnp">Previous SNP</a>
          <a class="btn btn-secondary jumpto" href="/nextsnp">Next SNP</a>
        </div>
      </div>
      <div class="col jumpto">
        <p></p>
        <p>Jump to:</p>
        <a href="#sub-freq">Substitutions and frequencies</a><br>
        <a href="#bind-affinity">Change in binding affinity (∆∆G)</a><br>
        {% if pdb != 'N/A' %}
        <a href="#protein-viewer">Protein viewer</a><br>
        {% endif %}
      </div>
    </div>

    <h2 class="result-titles">{{snp}}</h2>
    <hr>
    <div class="row">
      <div class="col-sm">
        <table>
          <tr>
            <th class="info">Gene:</th>
            <td>{{gene}}</td>
          </tr>
          <tr>
            <th class="info">Chromosome:</th>
            <td>{{chr}}</td>
          </tr>
          <tr>
            <th class="info">Genotype:</th>
            <td>{{genotype}}</td>
          </tr>
        </table>
      </div>
      <div class="col-sm">
        <table>
          <tr>
            <th class="info" style="vertical-align: top;">Clinical significance:</th>
            <td>{% for i in clin %}
              {{i}} <br>
              {% endfor %}</td>
          </tr>
        </table>
      </div>
    </div>
    <br>
    <div class="row">
      <div class="col-sm">
      {% if len(condition) != 0 %}
      <strong>Related conditions:</strong>
      {% for i in range(0, len(condition)) %}{% if i == len(condition) - 1 %}{{condition[i]}}.{% else %}{{condition[i]}},
      {% endif %}
      {% endfor %}
      {% endif %}
    </div>
    </div>


    <h2 class="result-titles" id="sub-freq">Substitutions and frequencies</h2>
    <hr>
    <p class="grey-border">These polar area charts (which are similar to a pie chart but shows fractions in terms of
      area rather than angle) represent the nucleotide and/or amino acid substiutions that are documented for this SNP.
      The larger
      values indicate the subsitutions that are more commonly associated with the given SNP.</p>
    <div class="left-box"><canvas id="chart1" class="left-graph" width="20" height="10"></canvas></div>
    <div class="right-box"><canvas id="chart2" class="right-graph" width="20" height="10"></canvas></div>
    <p>The nucleotide substitutions (with frequency) are
      {% for i in range(0, len(sorted_nuclist[0])) %}
      {% if i == len(sorted_nuclist[0]) - 1 %}
      {{sorted_nuclist[0][i]}} ({{sorted_nuclist[1][i]}}).
      {% else %}
      {{sorted_nuclist[0][i]}} ({{sorted_nuclist[1][i]}}),
      {% endif %}
      {% endfor %}
      The amino acid substitutions (with frequency) are
      {% for i in range(0, len(sorted_aalist[0])) %}
      {% if i == len(sorted_nuclist[0]) - 1 %}
      {{sorted_aalist[0][i]}} ({{sorted_aalist[1][i]}}).
      {% else %}
      {{sorted_aalist[0][i]}} ({{sorted_aalist[1][i]}}),
      {% endif %}
      {% endfor %}</p>
    {% if freq1000g != 'N/A' and freqhapmap != 'N/A' %}
    <p class="grey-border">These pie charts represent the frequency of the nucleotide corresponding to the SNP
      compared to the frequency
      of all other nucleotides. The smaller percentage indicates rarer SNPs.</p>
    <div class="left-box"><canvas id="pie1" width="20" height="10" class="left-graph"></canvas></div>
    <div class="right-box"><canvas id="pie2" width="20" height="10" class="right-graph"></canvas></div>
    <p>From the 1000 Genomes Project, the {{ freq1000g[0] }} nuclotide has a frequency of {{freq1000g[3:]}}
      and from the HapMap Project, the {{ freqhapmap[0] }} nuclotide has a frequency of {{ freqhapmap[3:] }}.</p>
    {% else %}
    <p>There are no nucleotide frequencies for this SNP from the 1000 Genomes Project or the HapMap Project.</p>
    {% endif %}
    <script>
      var chart1 = new Chart(document.getElementById("chart1"), {
        type: 'polarArea',
        data: {
          labels: '{{sorted_nuclist[0]| tojson}}'.slice(1, -1).split(', '),
          datasets: [{
            backgroundColor: ["#351f39", "#726a95", "#719fb0", "#a0c1b8", "#c7cfb7", "#9dad7f", "#557174",
              "#889bae"
            ],
            data: '{{sorted_nuclist[1]| tojson}}'.slice(1, -1).split(', ')
          }]
        },
        options: {
          title: {
            display: true,
            text: 'Nucleotide Substitutions'
          },
          legend: {
            position: 'left'
          }
        }
      });
      var aa_val = '{{sorted_aalist[1]| tojson}}'.slice(1, -1).split(', ')
      var aa_label = '{{sorted_aalist[0]| tojson}}'.slice(1, -1).split(', ')
      if ('{{aa1}}' == 'N/A') {
        aa_val = [0];
        aa_label = ['N/A'];
      }
      var chart2 = new Chart(document.getElementById("chart2"), {
        type: 'polarArea',
        data: {
          labels: aa_label,
          datasets: [{
            backgroundColor: ["#351f39", "#726a95", "#719fb0", "#a0c1b8", "#c7cfb7", "#9dad7f", "#557174",
              "#889bae"
            ],
            data: aa_val
          }]
        },
        options: {
          title: {
            display: true,
            text: 'Amino acid Substitutions'
          },
          legend: {
            position: 'right'
          }
        }
      });
      var fraction1 = parseFloat('{{ freq1000g[3:] }}')
      var fraction2 = parseFloat('{{ freqhapmap[3:] }}')
      var colors = ['#5d1451', '#8D5A85'];
      var pie1 = new Chart(document.getElementById("pie1"), {
        type: 'pie',
        data: {
          labels: ['{{ freq1000g[0] }}', "Other"],
          datasets: [{
            backgroundColor: colors,
            data: [fraction1, 1 - fraction1]
          }]
        },
        options: {
          title: {
            display: true,
            text: '1000 Genomes Project'
          },
          legend: {
            position: 'left'
          }
        }
      });
      var pie2 = new Chart(document.getElementById("pie2"), {
        type: 'pie',
        data: {
          labels: ['{{ freqhapmap[0] }}', "Other"],
          datasets: [{
            backgroundColor: colors,
            data: [fraction2, 1 - fraction2]
          }]
        },
        options: {
          title: {
            display: true,
            text: 'HapMap Project'
          },
          legend: {
            position: 'right'
          }
        }
      });
    </script>

    <h2 class="result-titles" id="bind-affinity">Change in binding affinity (∆∆G)</h2>
    <hr>
    {% if aa1 != 'N/A' %}
    <p class="grey-border">Press the "Calculate ∆∆G" button to calculate the ∆∆G value using each predictor.</p>
    <form action="#" method="post">
      <!-- <input type="submit" name="ddgcalc" value="wild" onclick="$('#loading').show();"> -->
      <input type="submit" class="btn btn-primary" name="ddgcalc" value="Calculate ∆∆G" onclick="$('#loading').show();">
    </form>
    <div id="loading">
      <p>Loading ... </p>
      <!-- <img src="static/loadingimage.gif" alt="loading"> -->
    </div>
    {% else %}
    <p class="grey-border">The ∆∆G value cannot be calculated as there is no known amino acid substitution. </p>
    {% endif %}
    {% if ddgresults %}
    <br>
    <table id="results" class="display">
      <caption>∆∆G values calculated with {{pdbselect}} and {{aa1}}</caption>
      <tr>
        <th id="program">Program</th>
        <th id="deltadeltaG">∆∆G (kcal/mol)</th>
        <th id="effect">Effect</th>
      </tr>
      {% for i in range(0, len(ddgresults)-1) %}
      <tr>
        <td headers="program {{ddgresults[i][0]}}">{{ddgresults[i][0]}}</td>
        <td headers="deltadeltaG {{ddgresults[i][0]}}">{{ddgresults[i][1]}}</td>
        <td headers="effect {{ddgresults[i][0]}}">{{ddgresults[i][2]}}</td>
      </tr>
      {% endfor %}
      <tr>
        <th id="conclusion">Stability Consensus</th>
        <th id="space">-</th>
        <th id="consensus">{{ddgresults[len(ddgresults)-1][2]}}</th>
      </tr>
    </table>
    <br>
    <p>Note: UEP only evaluates mutations located in highly-packed interface positions. SAAMBE-3D and I-Mutant2.0
      Structure
      only calcuate ∆∆G if the position is present on the pdb structure.</p>
    {% endif %}
    {% if pdb != 'N/A' %}

    <h2 class="result-titles" id="protein-viewer">Protein viewer </h2>
    <hr>
    {% if ddgresults %}
    <p class="grey-border">The {{pdbselect}} structure. The {{gene}} protein is shown in purple. Selecting another PDB
      structure will recalculate ∆∆G.</p>
    <div>
      <form action="#" method="post">
        {% for i in pdb %}
        <input type="submit" class="btn btn-outline-dark" name="pdbselect" value={{i}}>
        {% endfor %}
      </form>
      <script src="https://cdn.rawgit.com/arose/ngl/v0.10.4-1/dist/ngl.js "></script>
      <script>
        var pdbfile = '{{ pdbselect }}';
        var rscb = "rcsb://";
        var input = rscb.concat(pdbfile);
        var chain = ':{{chain}}';
        document.addEventListener("DOMContentLoaded", function () {
          var stage = new NGL.Stage("viewport", {
            backgroundColor: 'white'
          });
          var schemeId = NGL.ColormakerRegistry.addSelectionScheme([
            ["purple", chain],
            ["grey", "*"]
          ])
          stage.loadFile(input).then(function (o) {
            o.addRepresentation("ribbon", {
              colorScheme: schemeId
            });
            o.autoView();
          });
        });
      </script>
      <div id="viewport"></div>
    </div>
    {% else %}
    <p class="grey-border">Calculate ∆∆G to view protein structure.</p>
    {% endif %}
    {% endif %}
  </div>
</body>
{% endblock %}